src/main/scala/karazin/scala/users/group/week2/homework/adt.scala

```diff
@@ -9,66 +9,53 @@ package karazin.scala.users.group.week2.homework
   * https://docs.scala-lang.org/scala3/book/types-adts-gadts.html
 */
 
+
 object adt:
-  
+
   enum ErrorOr[+V]:
-    
-    // Added to make it compilable. Remove it.
-    case DummyCase
-    
-    /* 
-      Two case must be defined: 
-      * a case for a regular value
-      * a case for an error (it should contain an actual throwable)
-     */
-  
-    /* 
-      The method is used for defining execution pipelines
-      Provide a type parameter, an argument and a result type
-      
-      Make sure that in case of failing the method with exception
-      no exception is thrown but the case for an error is returned
-    */ 
-    def flatMap = ???
 
-    /* 
-      The method is used for changing the internal object
-      Provide a type parameter, an argument and a result type
-      
-      Make sure that in case of failing the method with exception
-      no exception is thrown but the case for an error is returned
-     */
-    def map = ???
-  
-    /* 
-      The method is used for filtering
-      Provide a type parameter, an argument and a result type
-      
-      Make sure that in case of failing the method with exception
-      no exception is thrown but the case for an error is returned
-     */
-    def withFilter = ???
-  
-    /* 
-      The method is used for getting rid of internal box
-      Provide a type parameter, an argument and a result type
-    */
-    def flatten = ???
-    
-    /* 
-      The method is used for applying side effects without returning any result
-      Provide a type parameter, an argument and a result type
-    */
-    def foreach = ???
-      
+    case Or(x: V) extends ErrorOr[V]
+    case Error(ex: Throwable) extends ErrorOr[V]
+
+    // The method is used for defining execution pipelines
+    def flatMap[Q](f: V => ErrorOr[Q]): ErrorOr[Q] =
+      this match
+        case ErrorOr.Error(ex)  => ErrorOr.Error(ex)
+        case ErrorOr.Or(x)      => try f(x) catch {
+          case ex: Throwable    => ErrorOr.Error(ex)
+        }
+
+    //The method is used for changing the internal object
+    def map[Q](f: V => Q): ErrorOr[Q] =
+      this match
+        case ErrorOr.Error(ex) => ErrorOr.Error(ex)
+        case ErrorOr.Or(x)     => try ErrorOr.Or(f(x)) catch {
+          case ex: Throwable   => ErrorOr.Error(ex)
+        }
+
+    //The method is used for filtering
+    def withFilter(p: V => Boolean): ErrorOr[V] =
+      this match 
+        case ErrorOr.Error(ex) => ErrorOr.Error(ex)
+        case ErrorOr.Or(x) if p(x) => ErrorOr.Or(x)
```

Comment:
This method will fail with an exception in case of `p(x)` is false, because you haven't handled this case.
But it's really tricky. 
Add a comment or a code example if you have an idea how to fix that.